crates/shared/src/event_handling.rs

```diff
@@ -69,51 +83,139 @@ where
+            == (last_handled_block_number, last_handled_block_hash)
+        {
+            return Ok(EventRange {
+                history_range: None,
+                latest_blocks: vec![],
+                is_reorg: false,
+            });
+        }
+
+        // full range of blocks which are considered for event update
+        let block_range = RangeInclusive::new(
+            last_handled_block_number.saturating_sub(MAX_REORG_BLOCK_COUNT),
+            current_block_number,
+        );
+        ensure!(
+            !block_range.is_empty(),
+            "current block number according to node is {} which is more than {} blocks in the \
                  past compared to last handled block {}",
-                current_block, MAX_REORG_BLOCK_COUNT, last_handled_block
-            )
+            block_range.end(),
+            MAX_REORG_BLOCK_COUNT,
+            block_range.start()
         );
-        Ok(BlockNumber::Specific(from_block)..=BlockNumber::Latest(current_block))
+
+        let (history_range, latest_range) = split_range(block_range)?;
+        tracing::debug!(
+            "history range {:?}, latest_range {:?}",
+            history_range,
+            latest_range
+        );
+
+        let latest_blocks = self.block_retriever.blocks(latest_range).await?;
+        tracing::debug!(
+            "latest blocks: {:?} - {:?}",
+            latest_blocks.first(),
+            latest_blocks.last(),
+        );
+
+        // do not try to shorten the latest_blocks list if history range exists
+        // if history range exists then we want to update for the full range of blocks,
+        // otherwise history_blocks update would erase all subsequent blocks and we might have a gap in storage
+        let (latest_blocks, is_reorg) = match history_range {
+            Some(_) => (latest_blocks, true),
+            None => {
+                let (latest_blocks, is_reorg) = detect_reorg_path(&handled_blocks, &latest_blocks);
+                (latest_blocks.to_vec(), is_reorg)
+            }
+        };
+
+        tracing::debug!(
+            "final latest blocks {:?} - {:?}, is reorg: {}",
+            latest_blocks.first(),
+            latest_blocks.last(),
+            is_reorg
+        );
+
+        Ok(EventRange {
+            history_range,
+            latest_blocks,
+            is_reorg,
+        })
     }
 
     /// Get new events from the contract and insert them into the database.
     pub async fn update_events(&mut self) -> Result<()> {
-        let range = self.event_block_range().await?;
-        tracing::debug!("updating events in block range {:?}", range);
+        let event_range = self.event_block_range().await?;
+
+        if let Some(range) = event_range.history_range {
+            self.update_events_from_old_blocks(range).await?;
+        }
+        self.update_events_from_latest_blocks(&event_range.latest_blocks, event_range.is_reorg)
+            .await?;
```

Comment:
One thing that is easy to get wrong here and in the inner functions is what happens when part of the function errors. Will the everything still work out of we only update old blocks and new blocks fail? What happens if we insert events into the database but then getting the hashes we need to update last_handled_blocks fails. So just wanted to make sure you thought about this.

Reply:
I _think_ if history update (which is always followed with `is_reorg = true`) fails, then in the next iteration is guaranteed to be again `is_reorg = true` (with or without history update) which triggers `REPLACE` instead of `APPEND` and that would fix the problem. But why bother with this, if we can move this failable part of the code to the beginning of the function and be certain.

Reply:
Moved the failable code to the beginning of the function.